presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -382,14 +382,27 @@ private Expression unwrapCast(ComparisonExpression expression)
             return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
         }
 
-        private boolean hasInjectiveImplicitCoercion(Type source, Type target)
+        private boolean hasInjectiveImplicitCoercion(Type source, Type target, Object value)
         {
             if ((source.equals(BIGINT) && target.equals(DOUBLE)) ||
                     (source.equals(BIGINT) && target.equals(REAL)) ||
                     (source.equals(INTEGER) && target.equals(REAL))) {
                 // Not every BIGINT fits in DOUBLE/REAL due to 64 bit vs 53-bit/23-bit mantissa. Similarly,
                 // not every INTEGER fits in a REAL (32-bit vs 23-bit mantissa)
-                return false;
+                if (target.equals(DOUBLE)) {
+                    double doubleValue = (double) value;
+                    return doubleValue > Long.MAX_VALUE ||
+                            doubleValue < Long.MIN_VALUE ||
+                            Double.isNaN(doubleValue) ||
```

Comment:
Can you explain why we return true in these cases?

Reply:
For cases where the double literal is greater (or less) than any possible bigint value when casted to a double, we can treat it as a beyond/below type boundary check for bigint type - example if the expression is like `bigint_col < cast(2^100 as double)`, then we can unwrap that case.
The NaN case is sent true because it gets unwrapped later as a part of NaN unwrap opt. done recently.